pycyphal/transport/udp/_frame.py

```diff
@@ -66,24 +120,126 @@ def compile_header_and_payload(self) -> typing.Tuple[memoryview, memoryview]:
         The reason is to avoid unnecessary data copying in the user space,
         allowing the caller to rely on the vectorized IO API instead (sendmsg).
         """
-        header = self._HEADER_FORMAT.pack(
-            self._VERSION, int(self.priority), self.index | ((1 << 31) if self.end_of_transfer else 0), self.transfer_id
-        )
+
+        # compute snm, subject_id, service_id, rnr (based on data_specifier)
+        # snm: bool  # Service, Not Message
+        # subject_id: int | None
+        # service_id: int | None
+        # rnr: bool | None  # Request, Not Response
+        snm = isinstance(self.data_specifier, pycyphal.transport.ServiceDataSpecifier)
+        subject_id = self.data_specifier.subject_id if not snm else None
+        service_id = self.data_specifier.service_id if snm else None
+        rnr = self.data_specifier.role == self.data_specifier.Role.REQUEST if snm else None
+        if snm:
+            id_rnr = service_id | ((1 << 14) if rnr else 0)
+        else:
+            id_rnr = subject_id
+
+        if self.end_of_transfer:
+            header_crc = 0
+            header_memory = self._HEADER_FORMAT_NO_CRC.pack(
+                self._VERSION,
+                int(self.priority),
+                self.source_node_id if self.source_node_id is not None else 0xFFFF,
+                self.destination_node_id if self.destination_node_id is not None else 0xFFFF,
+                ((1 << 15) if snm else 0) | id_rnr,
+                self.transfer_id,
+                ((1 << 31) if self.end_of_transfer else 0) | self.index,
+                0,  # user_data
+            )
+            crc = pycyphal.transport.commons.crc.CRC16CCITT()
+            crc.add(header_memory)
+            header_crc = crc.value
+            header = self._HEADER_FORMAT.pack(
+                self._VERSION,
+                int(self.priority),
+                self.source_node_id if self.source_node_id is not None else 0xFFFF,
+                self.destination_node_id if self.destination_node_id is not None else 0xFFFF,
+                ((1 << 15) if snm else 0) | id_rnr,
+                self.transfer_id,
+                ((1 << 31) if self.end_of_transfer else 0) | self.index,
+                0,  # user_data
+                header_crc,
+            )
```

Comment:
Instead of recompiling the header again why don't we simply go like `header_memory + header_crc.to_bytes(2, 'little')`